pkg/deploy/depoy.go

```diff
@@ -0,0 +1,247 @@
+package deploy
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	azcompute "github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-03-01/compute"
+	azresources "github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2018-05-01/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/resources"
+	"github.com/Azure/ARO-RP/pkg/util/ready"
+)
+
+var _ Deployer = (*deployer)(nil)
+
+type Deployer interface {
+	PreDeploy(context.Context, *logrus.Entry) (string, error)
+	Deploy(context.Context, *logrus.Entry, string) error
+	Upgrade(context.Context, *logrus.Entry) error
+}
+
+type deployer struct {
+	log         *logrus.Entry
+	deployments resources.DeploymentsClient
+	vmss        compute.VirtualMachineScaleSetsClient
+	vmssvms     compute.VirtualMachineScaleSetVMsClient
+	network     network.PublicIPAddressesClient
+
+	cli *http.Client
+
+	parameters        map[string]interface{}
+	version           string
+	resourceGroupName string
+	subscriptionID    string
+	location          string
+}
+
+// New initiates new deploy utility object
+func New(ctx context.Context, log *logrus.Entry, authorizer autorest.Authorizer, gitCommit string) (*deployer, error) {
+	d := &deployer{
+		log:               log,
+		version:           gitCommit,
+		resourceGroupName: os.Getenv("AZURE_RP_RESOURCEGROUP_NAME"),
+		subscriptionID:    os.Getenv("AZURE_SUBSCRIPTION_ID"),
+	}
+
+	d.cli = &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	// extract location from rg
+	gCli := resources.NewGroupsClient(d.subscriptionID, authorizer)
+	group, err := gCli.Get(ctx, d.resourceGroupName)
+	if err != nil {
+		return nil, err
+	}
+	d.location = *group.Location
+
+	d.deployments = resources.NewDeploymentsClient(d.subscriptionID, authorizer)
+	d.vmss = compute.NewVirtualMachineScaleSetsClient(d.subscriptionID, authorizer)
+	d.vmssvms = compute.NewVirtualMachineScaleSetVMsClient(d.subscriptionID, authorizer)
+	d.network = network.NewPublicIPAddressesClient(d.subscriptionID, authorizer)
+
+	return d, nil
+}
+
+// PreDeploy deploys NSG and ManagedIdentity, needed for man deployment
+func (d *deployer) PreDeploy(ctx context.Context, log *logrus.Entry) (rpServicePrincipalID string, err error) {
+	// if RP_PREDEPLOY is not set, we read old rpServicePrincipalID and return it
+	if len(os.Getenv("AZURE_RP_PREDEPLOY")) != 0 {
```

Comment:
don't understand the rationale for this env var - you already test to see if this has been deployed, why won't that do instead of the env var?
if someone decides they want to redeploy the nsg, they can delete the deployment manually and re-run